eth/filters/filter_system.go

```diff
@@ -14,179 +14,356 @@
-		pendingLogFilters: make(map[int]*Filter),
-		generic:           make(map[int]*Filter),
+// UnmarshalJSON parses a SubscriptionID from its JSON representation.
+func (f *ID) UnmarshalJSON(data []byte) error {
+	// `"0x...."`
+	if len(data) != 36 || data[0] != '"' || data[35] != '"' || data[1] != '0' || (data[2] != 'x' && data[2] != 'X') {
+		return ErrInvalidSubscriptionID
 	}
-	fs.sub = mux.Subscribe(
-		core.PendingLogsEvent{},
-		core.RemovedLogsEvent{},
-		core.ChainEvent{},
-		core.TxPreEvent{},
-		vm.Logs(nil),
-	)
-	go fs.filterLoop()
-	return fs
+
+	_, err := hex.Decode(f[:], data[3:35])
+	return err
 }
 
-// Stop quits the filter loop required for polling events
-func (fs *FilterSystem) Stop() {
-	fs.sub.Unsubscribe()
+// Log is a helper that can hold additional information about vm.Log
+// necessary for the RPC interface.
+type Log struct {
+	*vm.Log
+	Removed bool `json:"removed"`
 }
 
-// Acquire filter system maps lock, required to force lock acquisition
-// sequence with filterMu acquired first to avoid deadlocks by callbacks
-func (fs *FilterSystem) Lock() {
-	fs.filterMu.Lock()
+func (l *Log) MarshalJSON() ([]byte, error) {
+	fields := map[string]interface{}{
+		"address":          l.Address,
+		"data":             fmt.Sprintf("%#x", l.Data),
+		"blockNumber":      fmt.Sprintf("%#x", l.BlockNumber),
+		"logIndex":         fmt.Sprintf("%#x", l.Index),
+		"blockHash":        l.BlockHash,
+		"transactionHash":  l.TxHash,
+		"transactionIndex": fmt.Sprintf("%#x", l.TxIndex),
+		"topics":           l.Topics,
+		"removed":          l.Removed,
+	}
+
+	return json.Marshal(fields)
 }
 
-// Release filter system maps lock
-func (fs *FilterSystem) Unlock() {
-	fs.filterMu.Unlock()
+// subscriptionIDGenerator helper utility that generates a (pseudo) random sequence of bytes
+// that are used to generate filter identifiers.
+func subscriptionIDGenerator() *rand.Rand {
+	if seed, err := binary.ReadVarint(bufio.NewReader(crand.Reader)); err == nil {
+		return rand.New(rand.NewSource(seed))
+	}
+	return rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
 }
 
-// Add adds a filter to the filter manager
-// Expects filterMu to be locked.
-func (fs *FilterSystem) Add(filter *Filter, filterType FilterType) (int, error) {
-	id := fs.filterId
-	filter.created = time.Now()
+// newID generates a subscription identifier.
+func newID() ID {
+	filterIDGenMu.Lock()
+	defer filterIDGenMu.Unlock()
 
-	switch filterType {
-	case ChainFilter:
-		fs.chainFilters[id] = filter
-	case PendingTxFilter:
-		fs.pendingTxFilters[id] = filter
-	case LogFilter:
-		fs.logFilters[id] = filter
-	case PendingLogFilter:
-		fs.pendingLogFilters[id] = filter
-	default:
-		return 0, fmt.Errorf("unknown filter type %v", filterType)
+	var id ID
+	for i := 0; i < len(id); i += 7 {
+		val := filterIDGen.Int63()
+		for j := 0; i+j < len(id) && j < 7; j++ {
+			id[i+j] = byte(val)
+			val >>= 8
+		}
 	}
-	fs.generic[id] = filter
 
-	fs.filterId++
+	return id
+}
```

Comment:
Aren't this filterIDGen code and the subscriptionIDGen code in `rpc` equivalent? If so, wouldn't it make sense to move them into an internal package to avoid the copy paste as imho it's not "that" trivial and would be nice to have it in the code once only? :)